log4j-core/src/main/java/org/apache/logging/log4j/core/config/AbstractConfiguration.java

```diff
@@ -1178,4 +1152,28 @@ public NanoClock getNanoClock() {
     public void setNanoClock(final NanoClock nanoClock) {
         instanceFactory.registerBinding(NanoClock.KEY, () -> nanoClock);
     }
+
+    @Override
+    public void addExtension(ConfigurationExtension extension) {
+        Objects.requireNonNull(extension);
+        extensions = Arrays.copyOf(extensions, extensions.length + 1);
+        extensions[extensions.length - 1] = extension;
+    }
+
+    @Override
+    public <T extends ConfigurationExtension> T getExtension(Class<T> extensionType) {
+        T result = null;
+        for (final ConfigurationExtension extension : extensions) {
+            if (extensionType.isInstance(extension)) {
+                if (result == null) {
+                    result = (T) extension;
+                } else {
+                    LOGGER.warn(
+                            "Multiple configuration elements found for type {}. Only the first will be used.",
+                            extensionType.getName());
+                }
+            }
+        }
+        return result;
+    }
```

Comment:
My idea behind this is to prevent plugin implementors from abusing this feature by adding a lot of new elements to the `<Configuration>` tag.

Let's say an extension wants to add multiple `<Recycler>` elements to a configuration. Then I would strongly prefer to put them in a **single** `<Recyclers>` wrapper element, instead of polluting the main `<Configuration>` element.

Reply:
> My idea behind this is to prevent plugin implementors from abusing this feature by adding a lot of new elements to the `<Configuration>` tag.

I understand the concern. Though my point still stands: suppressing unexpected conditions is... unexpected, and, hence, not right. We can also address your concern by simply failing, which would be of my preference.

For the record, let's also agree that _"plugin implementors"_ is a handful of people around the globe.

Reply:
IMHO this unexpected condition should be dealt in a uniform way.

E.g. if a user adds multiple `<Disruptor>` elements all but one should be ignored and a warning should be issued. This is better than the logic for multiple `<Appenders>` elements, where all but one are ignored **without** a warning.